File: too-many-cooks/lib/src/tools/status_tool.dart
/// Status tool - system overview.
library;

import 'dart:convert' show jsonEncode;

import 'package:dart_logging/dart_logging.dart';
import 'package:dart_node_mcp/dart_node_mcp.dart';
import 'package:nadz/nadz.dart';
import 'package:too_many_cooks/src/data/data.dart';
import 'package:too_many_cooks/src/types.dart';

/// Input schema for status tool (no inputs required).
const statusInputSchema = <String, Object?>{
  'type': 'object',
  'properties': <String, Object?>{},
};

/// Tool config for status.
const statusToolConfig = (
  title: 'Status',
  description: 'Get system overview: agents, locks, plans, messages',
  inputSchema: statusInputSchema,
  outputSchema: null,
  annotations: null,
);

/// Create status tool handler.
ToolCallback createStatusHandler(TooManyCooksDb db, Logger logger) =>
    (args, meta) async {
      final log = logger.child({'tool': 'status'});

      final List<Map<String, Object?>> agents;
      switch (db.listAgents()) {
        case Success(:final value):
          agents = value.map(agentIdentityToJson).toList();
        case Error(:final error):
          return _errorResult(error);
      }

      final List<Map<String, Object?>> locks;
      switch (db.listLocks()) {
        case Success(:final value):
          locks = value.map(fileLockToJson).toList();
        case Error(:final error):
          return _errorResult(error);
      }

      final List<Map<String, Object?>> plans;
      switch (db.listPlans()) {
        case Success(:final value):
          plans = value.map(agentPlanToJson).toList();
        case Error(:final error):
          return _errorResult(error);
      }

      final List<Map<String, Object?>> messages;
      switch (db.listAllMessages()) {
        case Success(:final value):
          messages = value.map(messageToJson).toList();
        case Error(:final error):
          return _errorResult(error);
      }

      log.debug('Status queried');

      return (
        content: <Object>[
          textContent(
            jsonEncode({
              'agents': agents,
              'locks': locks,
              'plans': plans,
              'messages': messages,
            }),
          ),
        ],
        isError: false,
      );
    };

CallToolResult _errorResult(DbError e) => (
  content: <Object>[textContent(jsonEncode(dbErrorToJson(e)))],
  isError: true,
);

File: too-many-cooks/lib/src/tools/status_tool.ts
/// Status tool - system overview.

import type { Logger } from "../logger.js";
import {
  type TooManyCooksDb,
  agentIdentityToJson,
  agentPlanToJson,
  fileLockToJson,
  messageToJson,
} from "../data/data.js";
import {
  textContent,
  type CallToolResult,
  type ToolCallback,
} from "../types.js";
import { makeErrorResult } from "./tool_utils.js";

/** Input schema for status tool (no inputs required). */
export const STATUS_INPUT_SCHEMA = {
  type: "object",
  properties: {},
} as const;

/** Tool config for status. */
export const STATUS_TOOL_CONFIG = {
  title: "Status",
  description: "Get system overview: agents, locks, plans, messages",
  inputSchema: STATUS_INPUT_SCHEMA,
  outputSchema: null,
  annotations: null,
} as const;

/** Create status tool handler. */
export const createStatusHandler = (
  db: TooManyCooksDb,
  logger: Logger,
): ToolCallback =>
  async (): Promise<CallToolResult> => {
    const log = logger.child({ tool: "status" });

    const agentsResult = db.listAgents();
    if (!agentsResult.ok) {return await Promise.resolve(makeErrorResult(agentsResult.error));}
    const agents = agentsResult.value.map(agentIdentityToJson);

    const locksResult = db.listLocks();
    if (!locksResult.ok) {return await Promise.resolve(makeErrorResult(locksResult.error));}
    const locks = locksResult.value.map(fileLockToJson);

    const plansResult = db.listPlans();
    if (!plansResult.ok) {return await Promise.resolve(makeErrorResult(plansResult.error));}
    const plans = plansResult.value.map(agentPlanToJson);

    const messagesResult = db.listAllMessages();
    if (!messagesResult.ok) {return await Promise.resolve(makeErrorResult(messagesResult.error));}
    const messages = messagesResult.value.map(messageToJson);

    log.debug("Status queried");

    return await Promise.resolve({
      content: [
        textContent(JSON.stringify({ agents, locks, plans, messages })),
      ],
      isError: false,
    });
  };
